ament_cmake_python/cmake/ament_python_install_module.cmake

```
@@ -60,12 +60,14 @@ function(_ament_cmake_python_install_module module_file)
   get_filename_component(module_file "${module_file}" NAME)
   if(NOT ARG_SKIP_COMPILE)
     get_executable_path(python_interpreter Python3::Interpreter CONFIGURE)
+    set(_bytecompile_target "\$ENV{DESTDIR}${CMAKE_INSTALL_PREFIX}/${ARG_DESTINATION}/${module_file}")
```

There's no `${ARG_DESTINATION}`

```suggestion
    set(_bytecompile_target "\$ENV{DESTDIR}${CMAKE_INSTALL_PREFIX}/${destination}/${module_file}")
```

---

It looks like we get `DESTDIR` support if we use `install(FILES ... DESTINATION ...)` https://cmake.org/cmake/help/latest/command/install.html#common-options

IIUC currently this uses `execute_process` to byte-compile Python code directly into the install directory at configure time. I recommend making the output of `execute_process` go to the build directory, and then use `install()` to copy that file at build time. IIUC we don't need to use `DESTDIR` here anymore because `install()` would take care of that.

---

agreed! I just noticed this issue in my repository and made [this issue](https://github.com/ament/ament_cmake/issues/609)

Ideally, at configure-time, the install directory is not touched at all. My particular issue was that my install directory was owned by root (since I would `sudo make install`), so this invocation to python during configure-time when I rebuilt my project would cause a bunch of permission errors from the compileall.

Thank you for fixing!

---

The `execute_process` invocations here are part of an `install(CODE)` block, so they get run as a sort of CMake script during `make install`, not during configure time like a plain `execute_process`.

One downside to performing the bytecompile in the staging directory is that building with bytecompile ON and then again with it OFF will still install the old build's compiled artifacts unless we add some mechanism to handle tracking or cleaning up the artifacts.

---

Does it get executed as part of the `make package` target? Should it?

That was what I believe was happening with me. I would argue that it should only execute as part of the `make install` target but not the `make package` target, in case I am only executing the packaging step.

---

> Does it get executed as part of the `make package` target? Should it?

I'm not sure, but I would assume that the packaging process will involve staging the artifacts by installing them into a temporary directory (it certainly does for deb and RPM packages). The byte compilation would probably happen during that staging process.

We don't use `make package` in any of our supported workflows right now, so you'll have to give it a shot and report back.

---

OK, I tested your change in my environment, and it fixed my issue! CPack sets DESTDIR to the staging folder automatically, so this change now generates the compiled pyc files in the staging folder and brings it into the package automatically. 

This MR will resolve #609 

Thanks for the help!